src/coreclr/scripts/get_commit_hash.py

```diff
@@ -0,0 +1,170 @@
+#!/usr/bin/env python3
+#
+## Licensed to the .NET Foundation under one or more agreements.
+## The .NET Foundation licenses this file to you under the MIT license.
+#
+##
+# Title: get_commit_hash.py
+#
+# Notes:
+#
+# Script to store git commit hash of certain paths and store them in a file.
+# Used by CacheTask@2 task of yml pipeline.
+#
+################################################################################
+################################################################################
+
+import argparse
+import os
+from coreclr_arguments import *
+from jitutil import run_command
+
+parser = argparse.ArgumentParser(description="description")
+
+parser.add_argument("-configuration", help="Configuration text to identify the type of build. Might be OS/Arch, etc.")
+parser.add_argument("-run_type", help="Type of run. Options: 'coreclr', 'libraries'")
+parser.add_argument("-output", help="Path to output file")
+
+# Dictionary that maps the type of build to the file paths that build is affected by.
+# Commit hashes of all the paths for a specific builds are fetched and stored in an
+# output file. The output file is helpful in determining if there were any changes in
+# the paths and if a build is needed or not. If there were no changes, the cached build
+# (if available) will be used instead of doing a build.
+
+# 'common' is a special category. If anything is modified in one of those paths, we will
+# always run all the builds. So, in future, for diagnostic purpose, if the cache is hit in
+# scenarios where it was not supposed to hit, touching one of the paths listed under 'common'
+# will guarantee that nothing is restored from cache. This happens because when we fetch the
+# latest commit hash for those paths, it will always be a merge commit which will be different
+# than the previous recorded commits and hence will create a unique cache key which will not
+# be present in CacheTask.
+paths = {
+    'common': [
+        'global.json',
+        'Directory.Build.targets',
+        'Directory.Build.props',
+        'Directory.Solution.props',
+#        'eng'
+    ],
+    'coreclr': [
```

Comment:
This looks fragile. coreclr build depends on code under libraries (in particular, `libraries/Common` and `src/coreclr/System.Private.CoreLib`)

Reply:
Yes - As called out in "Lookouts" above, this piece is intentionally left open ended and I am expecting reviewers to point the right set of paths I should have under each category.

Reply:
I understand. Coming with the 100% right set of paths here is close to impossible.

Reply:
> provided the value we can get out of it.

FWIW: Some data from September 2021:
- The CI is triggered 2.36x times per PR on average 
- We spend ~150 hours of build machine time and ~300 hours of test machine time per PR on average
I do like this change and I think we should look into some sort of caching scheme like this. Where we are now, this change is only going to reduce the total CI costs by a few percent (<5%).

> it should be less risky and straightforward to enable this for test builds at least?

Yes, I think so.

Reply:
Another challenge that I faced, and I want to change is the following:

I tested this mechanics on internal pipelines where I trigged the run manually on the branch where I was developing this feature. So essentially, there were no changes to other folders/files except the ones in my branch. With that model, the cache was hit more frequently because the `git history` was linear. Only my changes were going into the PR branch.

When I ported it to public pipeline, I realized that on every single run, we do a MERGE commit that brings in newer changes to the PR branch and that invalidates the cache more often because whenever newer change is pushed on PR branch, when runs are triggered, newer changes can come from `main` branch, leading to not matching the cache often.